pyteal/ast/itxn.py

```diff
@@ -127,10 +145,23 @@ def SetField(cls, field: TxnField, value: Expr) -> Expr:
             value: The value to that the field should take. This must evaluate to a type that is
                 compatible with the field being set.
         """
-        return InnerTxnFieldExpr(field, value)
+        if not field.is_array:
+            if type(value) is list:
+                raise TealInputError(
+                    "inner transaction set field does not support array field"
+                )
+            return InnerTxnFieldExpr(field, cast(Expr, value))
+        else:
+            if type(value) is not list:
+                raise TealInputError(
+                    "inner transaction set array field does not support non-array field"
+                )
+            values = cast(List[Expr], value)
+            fieldsArrayToSet = [cls.SetField(field, valueIter) for valueIter in values]
```

Comment:
This line will call this function with a single element of the list and fail because the field is an array but the element is a single value